cellmonitor/source/android/app/src/main/java/org/kaaproject/kaa/demo/cellmonitor/manager/LocationManagerWrapper.java

```diff
@@ -20,58 +20,100 @@
-                continue;
-            }
-            if (bestLocation == null || location.getAccuracy() < bestLocation.getAccuracy()) {
-                bestLocation = location;
+        try {
+            for (String provider : mLocationManager.getAllProviders()) {
+                Location location = mLocationManager.getLastKnownLocation(provider);
+                if (location == null) {
+                    continue;
+                }
+                if (bestLocation == null || location.getAccuracy() < bestLocation.getAccuracy()) {
+                    bestLocation = location;
+                }
             }
-        }
 
-        if (bestLocation != null)
-            lastKnownLocation = bestLocation;
+            if (bestLocation != null)
+                lastKnownLocation = bestLocation;
+        } catch (SecurityException securityException) {
+            CellMonitorConstants.LOG.error("Can't receive last known location", securityException);
+            Toast.makeText(context, R.string.accept_location_permission, Toast.LENGTH_LONG).show();
+        }
 
         return lastKnownLocation;
     }
 
     public void pause() {
-        mLocationManager.removeUpdates(mGpsLocationListener);
+        try {
+            mLocationManager.removeUpdates(mGpsLocationListener);
+        } catch (SecurityException securityException) {
+            CellMonitorConstants.LOG.error("Can't pause location update");
+            throw securityException;
+        }
     }
 
     public void resume() {
         Criteria criteria = new Criteria();
-        String bestProvider = mLocationManager.getBestProvider(criteria, false);
-        mLocationManager.requestLocationUpdates(bestProvider, 0, 0, mGpsLocationListener);
+        criteria.setAccuracy(Criteria.ACCURACY_FINE);
+        String bestProvider = mLocationManager.getBestProvider(criteria, true);
+        try {
+            mLocationManager.requestLocationUpdates(bestProvider, 5, 1000, mGpsLocationListener);
+        } catch (SecurityException securityException) {
+            CellMonitorConstants.LOG.error("Can't resume location update", securityException);
+            throw securityException;
+        }
+
+        /*
+         * Force location update every 4 seconds util we receive the location
+         */
+        new Timer().schedule(new TimerTask() {
+            @Override
+            public void run() {
+                try {
+                    mGpsLocation = mLocationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
+
+                    if (mGpsLocation != null) {
+                        mGpsLocationListener.onLocationChanged(mGpsLocation);
+                        this.cancel();
+                    }
+                } catch (SecurityException securityException) {
+                    CellMonitorConstants.LOG.error("Can't get last known location", securityException);
+                    throw securityException;
+                }
+            }
+        }, 2000, 2000);
```

Comment:
Why four seconds? Why not in field? And there is no four seconds, read https://docs.oracle.com/javase/7/docs/api/java/util/Timer.html#schedule(java.util.TimerTask,%20java.util.Date,%20long)